src/libstd/panicking.rs

```diff
@@ -53,19 +55,43 @@ pub fn on_panic(obj: &(Any+Send), file: &'static str, line: u32) {
         (None, Some(ref mut err)) => {
             let _ = writeln!(err, "thread '{}' panicked at '{}', {}:{}",
                              name, msg, file, line);
-            if backtrace::log_enabled() {
-                let _ = backtrace::write(err);
-            }
+            let _ = backtrace::write(err);
         }
         _ => {}
     }
+}
 
-    // If this is a double panic, make sure that we printed a backtrace
-    // for this panic.
-    match err {
-        Some(ref mut err) if unwind::panicking() && !backtrace::log_enabled() => {
-            let _ = backtrace::write(err);
-        }
-        _ => {}
+pub fn on_panic(obj: &(Any+Send), file: &'static str, line: u32) {
+    let panics = PANIC_COUNT.with(|s| {
+        let count = s.get() + 1;
+        s.set(count);
+        count
+    });
+
+    // If this is the third nested call, on_panic triggered the last panic,
+    // otherwise the double-panic check would have aborted the process.
+    // Even if it is likely that on_panic was unable to log the backtrace,
+    // abort immediately to avoid infinite recursion, so that attaching a
+    // debugger provides a useable stacktrace.
+    if panics >= 3 {
+        util::dumb_print(format_args!("thread panicked while processing \
+                                       panic. aborting."));
+        unsafe { intrinsics::abort() }
+    }
+
+    // If this is a double panic, make sure that we print a backtrace
+    // for this panic. Otherwise only print it if logging is enabled.
+    if panics >= 2 || backtrace::log_enabled() {
+        log_panic(obj, file, line)
```

Comment:
I think that this wants to be called unconditionally in all cases actually, `backtrace::log_enabled` was actually just intended to guard the backtrace emission, not the panic message itself.

For example, if you have a panicking program with this patch and *don't* have `RUST_BACKTRACE=1`, this means that no output will happen. (e.g. see failing Travis build for an example)